backend/lambda.md

```diff
@@ -0,0 +1,41 @@
+---
+layout: page
+title: Running on AWS Lambda
+nav_order: 800
+parent: Backend
+---
+
+# Running the backend on AWS Lambda
+
+This page contains various notes on the deployment of the backend service on AWS Lambda.
+
+It is interesting to understand how lambdas work under the hood. In nutshell: a daemon makes a call on an [API endpoint](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html) (`/next`) which is blocking waiting for the next (HTTP) request to arrive. Then it passes the request to your lambda code with a timeout, and send the response back to that endpoint, and then wait back in loop on this `/next` endpoint. You do not know how many requests the daemon will process before being stopped and started on another server.
+
+
+## Running code after returning a response from an AWS Lambda function
+
+This may be useful to run some cleanup or some processing while responding quite quickly. The typical use case is propagations.
+
+[This interesting blog post from AWS](https://aws.amazon.com/blogs/compute/running-code-after-returning-a-response-from-an-aws-lambda-function/) describes several ways to achieve this goal.
+
+About the different options:
+- Invoking an async worker is a valid solution (used for propagation on June 2024) but still suffer to the delay to the async call. This is not really negligible, especially if we can encounter a cold start.
```

Comment:
Notes:
- The delay could be improved if instead of doing "service -> propagation endpoint -> propagation command" we do "service -> propagation command" (via https://docs.aws.amazon.com/lambda/latest/api/API_Invoke.html). But we would have to do something about concurrent calls. It might still be too slow though. How much delay are we talking about?
- I did some research for cold starts, that happens when a new Worker is created, apparently there are ways to mitigate the issue. Not sure if it's worth researching more.

Reply:
> The delay could be improved if instead of doing "service -> propagation endpoint -> propagation command" we do "service -> propagation command" (via https://docs.aws.amazon.com/lambda/latest/api/API_Invoke.html). But we would have to do something about concurrent calls. It might still be too slow though. 

I didn't want the backend code to be too AWS-specific. But you make me think it is trivial for me to the response streaming in Node, so the "Propagation endpoint" can answer immediately not waiting for the Invoke to complete.

> How much delay are we talking about?

The problem is more the worst cases where is could easy take 1sec I think.

> I did some research for cold starts, that happens when a new Worker is created, apparently there are ways to mitigate the issue. Not sure if it's worth researching more.

Yes I know there are solution, but the cold start problem is a bit inherent to the technology and cannot be solve in a cheap way if it is another lambda function as it is now.